File: sofia-core/src/main/java/com/antwerkz/sofia/Method.java
package com.antwerkz.sofia;

import java.text.DateFormat;
import java.text.Format;
import java.text.MessageFormat;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;

public class Method {
    private String value;
    private Boolean logged = Boolean.FALSE;
    private int argCount;
    private List<String> arguments = new ArrayList<String>();
    private List<String> parameters = new ArrayList<String>();
    private String key;
    private String logLevel;
    private String name;

    public Method(LoggingType type, String propKey, String value) {
        this.key = propKey;
        name = propKey;
        if (key.startsWith("@")) {
            logged = Boolean.TRUE;
            logLevel = key.substring(1, key.indexOf(".")).toLowerCase();
            name = key.substring(key.indexOf(".")+1);
        }
        this.value = value;
        countArguments(value);
    }

    public Boolean getLogged() {
        return logged;
    }

    public String getLogLevel() {
        return logLevel;
    }

    public String getKey() {
        return key;
    }

    public String getValue() {
        return value;
    }

    public int getArgCount() {
        return argCount;
    }

    public List<String> getArguments() {
        return arguments;
    }

    public List<String> getParameters() {
        return parameters;
    }

    private void countArguments(String value) {
        MessageFormat messageFormat = new MessageFormat(value);
        Format[] formats = messageFormat.getFormats();
        argCount = messageFormat.getFormats().length;
        for (int i = 0; i < argCount; i++) {
            parameters.add(String.format("%s arg%d", getType(formats[i]), i));
            arguments.add(String.format("arg%d", i));
        }
    }

    private String getType(Format format) {
        if (format == null) {
            return "Object";
        } else if (format instanceof DateFormat) {
            return "java.util.Date";
        } else if (format instanceof NumberFormat) {
            return "Number";
        }
        return "Object";
    }

    public String getMethodName() {
        final String[] parts = name.split("\\.");
        StringBuilder name = new StringBuilder();
        for (String part : parts) {
            name.append(name.length() != 0 ? LocalizerGenerator.capitalize(part) : part);
        }
        return name.toString();
    }

    public String getLoggerName() {
        final String[] parts = name.split("\\.");
        StringBuilder name = new StringBuilder();
        name.append("log");
        for (String part : parts) {
            name.append(Character.toTitleCase(part.charAt(0)))
                .append(part.substring(1));
        }
        return name.toString();
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder()
            .append("Method {")
            .append("key='").append(key).append('\'')
            .append(", parameters=").append(parameters)
            .append(", logged=").append(logged)
            .append(", logLevel='").append(logLevel).append('\'')
            .append(", value='").append(value).append('\'')
            .append(", argCount=").append(argCount)
            .append(", arguments=").append(arguments)
            .append('}');
        return sb.toString();
    }
}

File: sofia-core/src/main/kotlin/com/antwerkz/sofia/Method.kt
package com.antwerkz.sofia

import java.text.DateFormat
import java.text.Format
import java.text.MessageFormat
import java.text.NumberFormat
import java.util.ArrayList

class Method(val language: String, val loggingType: LoggingType, val key: String, val value: String) {
    var logged = java.lang.Boolean.FALSE
        private set
    var argCount: Int = 0
        private set
    val arguments = ArrayList<String>()
    val parameters = ArrayList<Pair<String, String>>()
    var logLevel: String? = null
        private set
    var name: String

    init {
        name = key
        if (key.startsWith("@")) {
            logged = java.lang.Boolean.TRUE
            logLevel = key.substring(1, key.indexOf(".")).toLowerCase()
            name = key.substring(key.indexOf(".") + 1)
        }
        countArguments()
    }

    private fun countArguments() {
        val messageFormat = MessageFormat(value)
        val formats = messageFormat.formats
        argCount = messageFormat.formats.size
        for (i in 0..argCount - 1) {
            parameters.add(Pair(getType(language, formats[i]), "arg${i}"))
            arguments.add("arg${i}")
        }
    }

    private fun getType(language: String, format: Format?): String {
        return when (format) {
            null -> if (language == "java") "Object" else "Any"
            is DateFormat -> "Date"
            is NumberFormat -> "Number"
            else -> "Object"
        }

    }

    val methodName: String
        get() {
            val parts = name.split("\\.".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
            val name = StringBuilder()
            for (part in parts) {
                name.append(if (name.length != 0) LocalizerGenerator.capitalize(part) else part)
            }
            return name.toString()
        }

    val loggerName: String
        get() {
            val parts = name.split("\\.".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
            val name = StringBuilder()
            name.append("log")
            for (part in parts) {
                name.append(Character.toTitleCase(part.get(0))).append(part.substring(1))
            }
            return name.toString()
        }

        override fun toString(): String {
            return "Method {key='${key}', parameters=${parameters}', logged=${logged}', logLevel='${logLevel}, value='${value}'," +
                    " argCount=${argCount}, arguments=${arguments}}"
        }
}